cpp/src/io/json/write_json.cu

```diff
@@ -418,53 +517,40 @@ struct column_to_strings_fn {
-      auto new_offsets = cudf::lists::detail::get_normalized_offsets(
-        lists_column_view(column), stream_, rmm::mr::get_current_device_resource());
-      auto const list_child_string = make_lists_column(
-        column.size(),
-        std::move(new_offsets),
-        std::move(child_string),
-        column.null_count(),
-        cudf::detail::copy_bitmask(column, stream_, rmm::mr::get_current_device_resource()),
-        stream_);
-      return strings::detail::join_list_elements(lists_column_view(*list_child_string),
-                                                 string_scalar{","},
-                                                 narep,
-                                                 strings::separator_on_nulls::YES,
-                                                 strings::output_if_empty_list::EMPTY_STRING,
-                                                 stream_,
-                                                 mr_);
-    }();
-    // create column with "[", "]" to wrap around list string
-    auto prepend = make_column_from_scalar(string_scalar{"["}, column.size(), stream_, mr_);
-    auto append  = make_column_from_scalar(string_scalar{"]"}, column.size(), stream_, mr_);
-    return cudf::strings::detail::concatenate(
-      table_view{{prepend->view(), list_string->view(), append->view()}},
-      string_scalar(""),
-      string_scalar("", false),
-      strings::separator_on_nulls::YES,
-      stream_,
-      mr_);
+
+    auto child_string_with_null = [&]() {
+      if (child_view.type().id() == type_id::STRUCT) {
+        return (*this).template operator()<cudf::struct_view>(
+          child_view,
+          children_names.size() > child_index ? children_names[child_index].children
+                                              : std::vector<column_name_info>{});
+      } else if (child_view.type().id() == type_id::LIST) {
+        return (*this).template operator()<cudf::list_view>(child_view,
+                                                            children_names.size() > child_index
+                                                              ? children_names[child_index].children
+                                                              : std::vector<column_name_info>{});
+      } else {
+        return cudf::type_dispatcher(child_view.type(), *this, child_view);
+      }
+    };
+    auto new_offsets = cudf::lists::detail::get_normalized_offsets(
+      lists_column_view(column), stream_, rmm::mr::get_current_device_resource());
+    auto const list_child_string = make_lists_column(
+      column.size(),
+      std::move(new_offsets),
+      std::move(child_string_with_null()),
+      column.null_count(),
+      cudf::detail::copy_bitmask(column, stream_, rmm::mr::get_current_device_resource()),
+      stream_);
+    auto l_pre = string_scalar{"["}, l_post = string_scalar{"]"}, sep = string_scalar{","},
+         elem_narep = string_scalar{"null"};
```

Comment:
could be constexpr

Reply:
`string_scalar` has non-trivial constructor. It is non-literal type, so it can't be a constexpr.